src/connection.rs

```diff
@@ -69,7 +69,14 @@ impl Connection {
 
             let ack: Packet<Ack> = self.socket.expect_packet(&buf[..bytes_recvd])?;
 
-            assert_eq!(Block::new(current_block), ack.body.block);
+            if Block::new(current_block) != ack.body.block {
+                let error = Packet::<Error>::error(
+                    Code::IllegalOperation,
+                    "Got an ack with the wrong packet number, terminate excecution",
+                );
```

Comment:
```suggestion
                let error = Packet::error(
                    Code::IllegalOperation,
                    format!("expected ACK for {} but got ACK for {}", current_block, ack.body.block),
                );
```

Awesome! Just a couple of tweaks. I don't think we have to turbofish `Packet::<Error>::error`, I think Rust knows which `Packet<T>` we mean when we use `Packet::error`.

I tweaked the error message to include more info.

I haven't tried compiling these suggestions so maybe a bit of massaging is necessary.